### PR TITLE
Retry asynchronous Aggregate Share responses from Helper 

### LINKED ISSUE #3931: DAP-15: Add asynchronous processing of Aggregate Shares
[This change](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/675) allows Aggregate Shares to be handled asynchronously. The Helper can choose to defer, giving an empty response as happens with AggregationJobResp and CollectionJobResp. The Leader must then obey the Retry-After semantics to complete the request.

### PR DESCRIPTION
**Pull request by jcjones**
This resolves  #3931. 

## Changes
- Changes the `CollectionJobDriver` to require a `ExponentialWithTotalDelayBuilder` rather than taking any `BackoffBuilder`, so we can call methods on it to limit lease times.
- Plumbs the `clock` through to `step_collection_job` so we can use it with `Retry-After` headers.
- When the leader encounters an empty body in response to an Aggregate Share request to a Helper, the leader parses the Retry-After (if any). Then, the collection job is released until the retry time, preferring the header value but falling back to the Retry Strategy delay.
- Adds `remaining_lease_duration` to `Lease`, which takes the place of the helper function `effective_lease_duration`  in the job driver.
- Changes the `MaybeLease` struct fields to `pub` so we can use them in tests, and uses them.

_Added 8 Aug_
-  Add the AggregateShareId type and plumb it through for the Leader
- Add the type to aggregate_share_jobs and collection_jobs tables
- Change the method for Aggregate Share requests from POST to PUT
- Change the Aggregate Share path to "{task-id}/aggregate_shares/{aggregate-share-id}"

> [!NOTE]
> This version does not implement changes to the `retry_http_request_notify` (see the backout in 676a32d74fbd9863e85423ea296c4b3dc906b0aa), but most of the legwork is done here. I will still add support for `DateTime`-style headers in #3943. 
>
> If we want to support automatic retries respecting the `Retry-After` header in `retry_http_request_notify` we still can, but I don't see much point.

## Test Comments

One thing the tests make clear is that we don't have any path to take a Retry-After value and make it known to the entity calling the `collection_job_driver`'s `step_collection_job` -- presumably the `JobDriver`. Perhaps that's a mechanism we want to add, passing along the Retry-After header when the job doesn't move forward?

### CODE REVIEW

File: aggregator_core/src/datastore/models.rs [hunk 1]
```diff
@@ -613,6 +614,18 @@ impl<T> Lease<T> {
     pub fn lease_attempts(&self) -> usize {
         self.lease_attempts
     }
+
+    /// Returns how long remains until the expiry time, with an optional clock skew allowance.
+    /// The math saturates, because  we want to timeout immediately if any of these
```
Comment (tgeoghegan) [inline aggregator_core/src/datastore/models.rs:619]:
```suggestion
    /// The math saturates, because we want to timeout immediately if any of these
```

File: aggregator_core/src/datastore/models.rs [hunk 2]
```diff
@@ -613,6 +614,18 @@ impl<T> Lease<T> {
...
+    /// subtractions would underflow.
+    pub fn remaining_lease_duration(&self, current_time: &Time, skew_seconds: u64) -> StdDuration {
```
Comment (tgeoghegan) [inline aggregator_core/src/datastore/models.rs:621]: Saturating math and depending on underflow for control logic is fine, but it always gives me pause. I think it'd be nice to write a test on this method both to make sure some idiot like me doesn't introduce regressions but also to express and document our intended semantics for leases, especially in edge cases like where `self.lease_expiry_time` is 0.

File: aggregator/src/aggregator/collection_job_driver.rs [hunk 3]
```diff
@@ -386,13 +388,43 @@ where
...
         .await?;
 
+        let response_headers = http_response.headers();
+        let response_body = http_response.body();
+        if response_body.is_empty() {
+            let retry_delay = match parse_retry_after_to_duration(response_headers) {
+                Ok(Some(retry_delay)) => retry_delay,
+                _ => self
+                    .collection_retry_strategy
+                    .clone()
```
Comment (tgeoghegan) [inline aggregator/src/aggregator/collection_job_driver.rs:398]: I'm surprised you need to `clone()` here since `RetryStrategy::compute_retry_delay` takes `&self`
https://github.com/divviup/janus/blob/6205834e3d4b281f506781b544b3f242aca815e2/aggregator/src/aggregator/collection_job_driver.rs#L790

File: aggregator/src/aggregator/collection_job_driver.rs [hunk 4]
```diff
@@ -1154,7 +1189,7 @@ mod tests {
         // Batch aggregations indicate not all aggregation jobs are complete, and there is an
         // unaggregated report in the interval.
         collection_job_driver
-            .step_collection_job(Arc::clone(&ds), Arc::clone(&lease))
+            .step_collection_job(Arc::clone(&ds), clock.clone(), Arc::clone(&lease))
```
Comment (tgeoghegan) [inline aggregator/src/aggregator/collection_job_driver.rs:1192]: Elsewhere, I think we make a point of using `Arc<C>` and cloning that, just in case a particular `Clock` impl happens to be big... but honestly, it's been years now, and the only `Clocks` we have are `MockClock` (which is just an `Arc<Mutex<Time>>` and so roughly pointer sized and already provides interior mutability) or `RealClock` which I think is an empty struct that just calls `Instant::now`. So  think we'd save ourselves some hassle throughout Janus by having all `Clock`s be `Copy`.

Anyway `clock.clone()` is fine for this PR, but I wanted to offer some context.

https://github.com/divviup/janus/issues/3948

File: aggregator/src/aggregator/collection_job_driver.rs [hunk 5]
```diff
@@ -1881,4 +1916,248 @@ mod tests {
...
+
+        // Simulate leader indicating asynchronous operation; e.g., for the first request,
+        // the share isn't yet ready, then it is.
+
+        // Helper aggregate share is opaque to the leader, so no need to construct a real one
+        let helper_response = AggregateShare::new(HpkeCiphertext::new(
+            HpkeConfigId::from(100),
+            Vec::new(),
+            Vec::new(),
+        ));
```
[the remark this answers is not in the document]
Comment (jcjones) [reply aggregator/src/aggregator/collection_job_driver.rs]: Added `fake_aggregate_share` to `aggregator/src/aggregator/test_util.rs`

File: aggregator/src/aggregator/collection_job_driver.rs (hunk 4 above)
Comment (tgeoghegan) [reply aggregator/src/aggregator/collection_job_driver.rs:1194]: Yeah, what's here is what we should continue to do. I was just wrong.

**Comment by tgeoghegan** [comment]
> Gonna have to have admin dismiss the [ci-migrations / janus_migrations (pull_request)](https://github.com/divviup/janus/actions/runs/16840462075/job/47710524227?pr=3947) status.

The way forward for this is to add the `allow-changed-migrations` GH label to the PR, and then force CI to re-run. The easiest thing is to tack on an empty commit or rewrite HEAD, either of which will cause GitHub to re-run pull requests triggered jobs.

**Review by tgeoghegan — COMMENTED** [review]
One more hurdle here, which is that the leader needs to poll on agg share state with GET and not PUT. But the retry logic is right, I think.

File: messages/src/lib.rs [hunk 6]
```diff
@@ -517,6 +517,123 @@ impl Distribution<ReportIdChecksum> for StandardUniform {
...
+#[derive(Copy, Clone, Debug, Default, Hash, PartialEq, Eq)]
+pub struct AggregateShareId([u8; Self::LEN]);
+
+impl AggregateShareId {
+    /// LEN is the length of an aggregate share ID in bytes.
+    pub const LEN: usize = 16;
+}
+
+impl From<[u8; Self::LEN]> for AggregateShareId {
+    fn from(checksum: [u8; Self::LEN]) -> Self {
```
[the remark this answers is not in the document]
Comment (jcjones) [reply messages/src/lib.rs:530]: oops

File: aggregator/src/aggregator/http_handlers.rs [hunk 7]
```diff
@@ -478,7 +480,7 @@ where
                 COLLECTION_JOB_ROUTE,
                 instrumented(api(collection_jobs_delete::<C>)),
             )
-            .post(
+            .put(
                 AGGREGATE_SHARES_ROUTE,
                 instrumented(api(aggregate_shares::<C>)),
             );
```
[the remark this answers is not in the document]
Comment (jcjones) [reply aggregator/src/aggregator/http_handlers.rs:486]: I meant to write that as a TODO, thanks for thinking of it, too. A later patch, indeed.

[the remark this answers is not in the document]
Comment (jcjones) [reply aggregator/src/aggregator/http_handlers.rs:486]: Actually, this has to happen at the same time as the transition from PUT to GET polling.

[the remark this answers is not in the document]
Comment (jcjones) [reply aggregator/src/aggregator/http_handlers.rs:486]: This got too big, so I'm splitting it off to https://github.com/divviup/janus/issues/3963.

File: aggregator/src/aggregator/collection_job_driver.rs [hunk 8]
```diff
@@ -386,13 +389,42 @@ where
...
+                Ok(Some(retry_delay)) => retry_delay,
+                _ => self
+                    .collection_retry_strategy
+                    .compute_retry_delay(lease.leased().step_attempts()),
+            };
+            info!(
+                ?retry_delay,
+                ?response_headers,
+                "Helper has deferred collection of the Aggregate Share request, releasing lease."
+            );
```
[the remark this answers is not in the document]
Comment (jcjones) [reply aggregator/src/aggregator/collection_job_driver.rs:405]: Handling in https://github.com/divviup/janus/issues/3963

File: messages/src/lib.rs [hunk 9]
```diff
@@ -517,6 +517,123 @@ impl Distribution<ReportIdChecksum> for StandardUniform {
...
+impl Distribution<AggregateShareId> for StandardUniform {
+    fn sample<R: rand::Rng + ?Sized>(&self, rng: &mut R) -> AggregateShareId {
+        AggregateShareId(rng.random())
+    }
+}
+
+/// This customized implementation serializes a [`TaskId`] as a base64url-encoded string, instead
+/// of as a byte array. This is more compact and ergonomic when serialized to YAML, and aligns with
+/// other uses of base64url encoding in DAP.
+impl Serialize for AggregateShareId {
```
[the remark this answers is not in the document]
Comment (jcjones) [reply messages/src/lib.rs]: Looking at it again, the `AggregateShareId`, as a url-safe 16-byte field, will always be 22 bytes long, so I'm now asserting that before decoding.

**Comment by jcjones** [comment]
whoops, didn't wait long enough to test the docker tests, and the e2e ones need an update

**Comment by jcjones** [comment]
oh, nevermind, it worked. And it works locally, too, so some spurious error.

**Comment by jcjones** [comment]
@tgeoghegan Ready for re-review. Note the two follow-ups: 
- https://github.com/divviup/janus/issues/3962
- https://github.com/divviup/janus/issues/3963

**Review by tgeoghegan — APPROVED** [review]
